src/NuGetGallery/Services/MarkdownService.cs

```
@@ -3,25 +3,58 @@
 
 using System;
 using System.IO;
+using System.Linq;
 using System.Text.RegularExpressions;
+using System.Timers;
 using System.Web;
 using CommonMark;
 using CommonMark.Syntax;
+using Markdig;
+using Markdig.Parsers;
+using Markdig.Renderers;
+using Markdig.Syntax;
+using Markdig.Syntax.Inlines;
 
 namespace NuGetGallery
 {
     public class MarkdownService : IMarkdownService
     {
         private static readonly TimeSpan RegexTimeout = TimeSpan.FromMinutes(1);
         private static readonly Regex EncodedBlockQuotePattern = new Regex("^ {0,3}&gt;", RegexOptions.Multiline, RegexTimeout);
-        private static readonly Regex CommonMarkLinkPattern = new Regex("<a href=([\"\']).*?\\1", RegexOptions.None, RegexTimeout);
+        private static readonly Regex LinkPattern = new Regex("<a href=([\"\']).*?\\1", RegexOptions.None, RegexTimeout);
+
+        private readonly IFeatureFlagService _features;
+
+        public MarkdownService(IFeatureFlagService features)
+        {
+            _features = features ?? throw new ArgumentNullException(nameof(features));
+        }
 
         public RenderedMarkdownResult GetHtmlFromMarkdown(string markdownString)
         {
```

Null check for a public method is needed.

---

? what eles attribute to check here? I believe features is only property we need check if is NULL?

---

We have checked "_features" in the constructor, so there is no need to check it again. But checking whether "markdownString" is null or not is needed. It's better to throw asap rather than wait until internal calls throw null exceptions.

---

Thanks for advice of null check here for `markdownString `, I think it might be good to return if  `markdownString` is null.  It's not necassary to throw a null exception here. it's posible the markdownString is null based on the readme sourceType. for example, customer put url as readme source, but there is nothing is readme. I don't think we want to throw an exception in this case. 
What do you think?

---

Personally, I prefer to keep throwing the exception. One reason is that we should align with the previous behavior: for "CommonMark", if we feed a null string, it will throw the exception because of `markdownString.StartsWith("\ufeff")`. So a better choice here is to follow this and avoid breaking the current caller's logic. Another reason is that this service will be widely shared among multiple components: licenses, change logs, and etc. We should determine a unified behavior of this compact and reliable service for the future usage, rather than just considering the readme design. It's the caller's responsibility to determine how to handle the exception or validate the input before feeding into this service, and I think readme services properly handle different cases before feeding into this service.

---

Added validation for parameter, please check latest changes